rfcs/2fa.md

```diff
@@ -0,0 +1,54 @@
+# 2FA
+
+The module provides possibility to add second factor to the authentication process.
+
+### Enabling 2FA
+
+If the user wants to enable 2FA service generates secret key and recovery code.
+The user may authorize secret key in any authenticator app (e.g Google Authenticator).
+From that point user has an ability to generate TOTP (time-based one time password).
+Service verifies that TOTP generated by the user is valid. If it is service attaches
+secret key and recovery code to the user's account.
+
+Recovery code may be used in the future to recover an access to an account in case
+user can't provide a valid TOTP (e.g. user lost an access to authenticator app).
+
+### Basic user flows
+
+If the user enabled 2FA the flow of authentication or performing some privileged
+action (e. g. payment transaction) would be like this:
+
+**Authentication**
+
+1. User performs `POST` request with credentials:
+    * **valid credentials**: proceed to next step
+    * **invalid credentials**: reject request
+2. Service checks that 2FA is enabled:
+    * **enabled**: ask user for TOTP
+    * **not enabled**: return `JWT`
+3. User get TOTP from mobile phone and passes it to service:
+    * **valid TOTP**: return `JWT`
+    * **invalid TOTP**: reject request
+
+**Password update**
+
+1. On this step we already know if 2FA is enabled:
```

Comment:
I don't think that we can rely on cached user data since 2FA obviously can be disabled from another device/browser, so lock/check should be applied before.

Reply:
What we can do is use special tokens to "sign" actions, so that we have a single protocol for all 2FA interactions
Ie, imagine the following scenario:

1. Any action is dispatched by the client (without any checks whether 2FA is enabled or not)
2. Server verifies if 2FA is enabled, and if so - creates an "action" token, which is basically a JWT with a single scope - `verify-totp`. After that server returns that token with some sort of error (412 or upgrade required - need to check standards so that we do not break anything), and possibly a link to verify the action - so that we can control this from the server - general output is a short-lived JWT + link to execute it
3. When frontend client gets such a response, it asks the user to provide second factor auth
4. It then submits that token & TOTP to the link from (2)
5. In case of success we do (200), in case of error - client can retry up to x times (or probably up to time of the JWT validity - maybe 5 minutes or 10 minutes). Possibly enforce limits per ip to avoid bruteforce